_posts/2024-04-20-websockets2.md

```diff
@@ -0,0 +1,1861 @@
+  def saveMessage(
+      message: String,
+      time: LocalDateTime,
+      userId: UserId,
+      roomId: RoomId
+  ): F[Unit]
+  def fetchMessages(roomId: RoomId): F[Stream[F, FetchMessage]]
+  def deleteRoomMessages(roomId: RoomId): F[Unit]
+}
+```
+First, we create a `PostgresProtocol` trait with a number of methods that will later be used by our chat protocol:
+
+```scala
+...
+import cats.effect.std.UUIDGen
+import cats.effect.* 
+import skunk.* 
+
+trait PostgresProtocol[F[_]] {
+  ...
+}
+
+object PostgresProtocol {
+  def make[F[_]: UUIDGen: MonadCancelThrow: Concurrent](
+      postgres: Resource[F, Session[F]]
+      ): PostgresProtocol[F] = 
+          new PostgresProtocol[F] {
+            ???
+          }
+}
+private object SqlCommands {
+  ???
+}
+```
+We also provide a companion object with a `make()` method where we'll implement all the methods defined in the trait, below that we have the `SqlCommands` object which will contain other `Codecs` used in our implementations.
+
+First we'll implement `createUser()`, which inserts a user into the database, to do this, we'll need a `Codec[User]` for this to work. 
+```scala
+import rockthejvm.websockets.codecs.codecs.*
+...
+private object SqlCommands {
+    val usercodec: Codec[User] = 
+        (userId ~ userName).imap {
+            case i ~ n => User(i,n)
+        }(u => (u.id, u.name))
+}
+```
+Here we create a `twiddle` list, `(userId ~ userName)` which produces a `Codec[(UserId, UserName)]`, we then call `imap` and provide the two functions from `userId ~ userName` to `User` and the reverse
+
+Let's also go ahead and implement the `Codec[Room]` since its very similar to `Codec[User]`:
+
+```scala
+private object SqlCommands {
+  ...
+  val roomcodec: Codec[Room] = 
+      (roomId ~ roomName).imap {
+          case i ~ n => Room(i,n)
+      }(u => (u.id, u.name))
+}
+```
+
+Next, we'll need to create its associate `Command` for the `INSERT` statement. In Skunk, `Select` statements are prepared with a `Query` while `Insert` and `Update` statements are prepared with a `Command` since it return no rows.
+
+```scala
+private object SqlCommands {
+  ...
+  val insertUser: Command[User] = 
+      sql"""
+          INSERT INTO users
+          VALUES($usercodec)
+      """
+      .command
+}
+```
+Here we prepare an insert statement, and pass the `usercodec` to `VALUES()`, then call the `command` method giving us a `Command[User]`. A `Command[User]`, means we need to pass a `User` inorder to execute the query:
+
+Let's also create `Command[Room]` since its also similar to the above:
+
+```scala
+private object SqlCommands {
+  ...
+  val insertRoom: Command[Room] =
+    sql"""
+            INSERT INTO rooms
+            VALUES($roomcodec)
+        """.command
+}
+```
+Now we can implement `createUser()`:
+
+```scala
+...
+new PostgresProtocol[F] {
+  override def createUser(name: String): F[Either[String, User]] =
+    postgres.use { session =>
```

Comment:
This instantiates the Resource every time. You'll need to pass the Session as an argument and instantiate the Resource just once.

Same for all other uses of `postgres.use`

Reply:
Updated